### DISCUSSION TITLE
How to compile it with zopfli instead of zlib?

### DISCUSSION THREAD

**Discussion by murty2** [qpdf/qpdf#1320] [Q&A]
I know that qpdf supports zlib compatible alternatives such as zopfli according to this link: https://github.com/qpdf/qpdf/issues/1036

I installed zopfli and zopfli-devel RPM packages on my Red Hat Ent Linux 9. I can't figure out the exact cmake command to build qpdf using zopfli instead of zlib. Details of my zopfli RPM package below. Can someone please help me with what cmake command I can use to build qpdf with zopfli?

```
$ rpm -ql zopfli-devel
/usr/include/zopfli.h
/usr/include/zopflipng_lib.h
/usr/lib64/cmake/Zopfli/ZopfliConfig-release.cmake
/usr/lib64/cmake/Zopfli/ZopfliConfig.cmake
/usr/lib64/cmake/Zopfli/ZopfliConfigVersion.cmake
/usr/lib64/libzopfli.so
/usr/lib64/libzopflipng.so
```
```
$ rpm -ql zopfli
/usr/bin/zopfli
/usr/bin/zopflipng
/usr/lib/.build-id
/usr/lib/.build-id/02
/usr/lib/.build-id/02/bc1242e95a482e3430ef09c38f5a08d0b78cd3
/usr/lib/.build-id/4e
/usr/lib/.build-id/4e/641d44341ec10de1e87d091e7e125796fc1429
/usr/lib/.build-id/b0
/usr/lib/.build-id/b0/1b5325ea0c374360f52b19375e959525159618
/usr/lib/.build-id/ef
/usr/lib/.build-id/ef/a8fd68cad0c73c4b8e0152ef543082c4f31b2f
/usr/lib64/libzopfli.so.1
/usr/lib64/libzopfli.so.1.0.3
/usr/lib64/libzopflipng.so.1
/usr/lib64/libzopflipng.so.1.0.3
/usr/share/doc/zopfli
/usr/share/doc/zopfli/CONTRIBUTORS
/usr/share/doc/zopfli/README
/usr/share/doc/zopfli/README.zopflipng
/usr/share/licenses/zopfli
/usr/share/licenses/zopfli/COPYING
```

**Comment by jberkenbilt** [comment]
Does zopfli have a drop-in zlib-compatible API replacement like zlibng does? If so, that would be your easiest path. If not, it's likely that code changes would be required. The only file that directly uses zlib is libqpdf/Pl_Flate.cc. It's also possible that there may need to be test suite changes, but I think I have the test suite wired up to be agnostic about zlib implementations. There's a section in README-maintainer.md about it, but it's been about a year and I don't remember the details without rereading it.

**Comment by jberkenbilt** [comment]
So it looks like there's nothing immediate that will do this with qpdf without some code changes, but they should be small. I'm going to convert this to an issue and add `next` to it. I'm hoping to find some time to get qpdf ready to release again in the next few weeks and maybe could add a cmake flag to use zopfli, unless someone beats me to it. The only reference to zlib.h is in Pl_Flate.cc -- it's not exposed in any public headers. The only reference to zlib for compile and link are in cmake, so I think it would be a quick fix.
